In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
data_path = "../data/notebooks/1_raw-data.csv"

In [3]:
df_raw = pd.read_csv(data_path)
df = df_raw.copy()

C:\Users\saima\Anaconda3\envs\kick\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (15,19,21,25) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [4]:
df =df_raw

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 215800 entries, 0 to 215799
Data columns (total 39 columns):
Unnamed: 0                  215800 non-null int64
backers_count               215800 non-null int64
blurb                       215792 non-null object
category                    215800 non-null object
converted_pledged_amount    215800 non-null int64
country                     215800 non-null object
country_displayable_name    215800 non-null object
created_at                  215800 non-null int64
creator                     215800 non-null object
currency                    215800 non-null object
currency_symbol             215800 non-null object
currency_trailing_code      215800 non-null bool
current_currency            215800 non-null object
deadline                    215800 non-null int64
disable_communication       215800 non-null bool
friends                     252 non-null object
fx_rate                     215800 non-null float64
goal                        21580

In [6]:
df[:5].transpose()

0  \
Unnamed: 0                                                                0   
backers_count                                                             2   
blurb                     Two Guys running a Boutique B&B in Sunny Spain...   
category                  {"id":10,"name":"Food","slug":"food","position...   
converted_pledged_amount                                                 54   
country                                                                  SE   
country_displayable_name                                             Sweden   
created_at                                                       1570742329   
creator                   {"id":1480510910,"name":"Peter Johansson","slu...   
currency                                                                SEK   
currency_symbol                                                          kr   
currency_trailing_code                                                 True   
current_currency                                                        USD   
deadline                                                         1576266603   
disable_communication                                                 False   
friends                                                                 NaN   
fx_rate                                                            0.106548   
goal                                                                 150000   
id                                                               2095605118   
is_backing                                                              NaN   
is_starrable                                                          False   
is_starred                                                              NaN   
launched_at                                                      1571079003   
location                  {"id":768372,"name":"Nerja","slug":"nerja-nerj...   
name                         Organic Garden @ Boutique Hotel in sunny Spain   
permissions                                                             NaN   
photo                     {"key":"assets/026/794/518/3c4061c549f67cbc846...   
pledged                                                              508.71   
profile                   {"id":3830311,"project_id":3830311,"state":"in...   
slug                           organic-garden-boutique-hotel-in-sunny-spain   
source_url                https://www.kickstarter.com/discover/categorie...   
spotlight                                                             False   
staff_pick                                                            False   
state                                                                  live   
state_changed_at                                                 1571079003   
static_usd_rate                                                    0.101984   
urls                      {"web":{"project":"https://www.kickstarter.com...   
usd_pledged                                                         51.8802   
usd_type                                                           domestic   

                                                                          1  \
Unnamed: 0                                                                1   
backers_count                                                            59   
blurb                     I need help! I'm recording my new record "Shad...   
category                  {"id":42,"name":"Pop","slug":"music/pop","posi...   
converted_pledged_amount                                               5818   
country                                                                  US   
country_displayable_name                                  the United States   
created_at                                                       1355429246   
creator                   {"id":93848347,"name":"Andrew Varner","slug":"...   
currency                                                                USD   
currency_symbol                             

- Let us first see what's up with these dictionary


In [7]:
df['category'][:10].values

array(['{"id":10,"name":"Food","slug":"food","position":8,"color":16725570,"urls":{"web":{"discover":"http://www.kickstarter.com/discover/categories/food"}}}',
       '{"id":42,"name":"Pop","slug":"music/pop","position":14,"parent_id":14,"color":10878931,"urls":{"web":{"discover":"http://www.kickstarter.com/discover/categories/music/pop"}}}',
       '{"id":54,"name":"Mixed Media","slug":"art/mixed media","position":6,"parent_id":1,"color":16760235,"urls":{"web":{"discover":"http://www.kickstarter.com/discover/categories/art/mixed%20media"}}}',
       '{"id":22,"name":"Illustration","slug":"art/illustration","position":4,"parent_id":1,"color":16760235,"urls":{"web":{"discover":"http://www.kickstarter.com/discover/categories/art/illustration"}}}',
       '{"id":342,"name":"Web","slug":"technology/web","position":15,"parent_id":16,"color":6526716,"urls":{"web":{"discover":"http://www.kickstarter.com/discover/categories/technology/web"}}}',
       '{"id":262,"name":"Accessories","slug":"fa

- Slug has the data in the format 'Category/Sub category'(Confirmed that by looking up categories and sub categories of a few campaigns


In [8]:
df['creator'][0]

'{"id":1480510910,"name":"Peter Johansson","slug":"organicfoods","is_registered":null,"chosen_currency":null,"is_superbacker":null,"avatar":{"thumb":"https://ksr-ugc.imgix.net/assets/026/794/760/80c439e676e400ef9d5b4aef742e7080_original.jpg?ixlib=rb-2.1.0&w=40&h=40&fit=crop&v=1570744152&auto=format&frame=1&q=92&s=9f05c652af592edce851251105cdd171","small":"https://ksr-ugc.imgix.net/assets/026/794/760/80c439e676e400ef9d5b4aef742e7080_original.jpg?ixlib=rb-2.1.0&w=160&h=160&fit=crop&v=1570744152&auto=format&frame=1&q=92&s=f352c58c6c887d4feb3a62f1a4fce303","medium":"https://ksr-ugc.imgix.net/assets/026/794/760/80c439e676e400ef9d5b4aef742e7080_original.jpg?ixlib=rb-2.1.0&w=160&h=160&fit=crop&v=1570744152&auto=format&frame=1&q=92&s=f352c58c6c887d4feb3a62f1a4fce303"},"urls":{"web":{"user":"https://www.kickstarter.com/profile/organicfoods"},"api":{"user":"https://api.kickstarter.com/v1/users/1480510910?signature=1576214609.1b03bc1684fc784282cd5fcb99d7d4b013e0d6d8"}}}'

- We've got the name of the creator in here

In [9]:
df['location'][0]

'{"id":768372,"name":"Nerja","slug":"nerja-nerja-es","short_name":"Nerja, Spain","displayable_name":"Nerja, Spain","localized_name":"Nerja","country":"ES","state":"Andalusia","type":"Town","is_root":false,"expanded_country":"Spain","urls":{"web":{"discover":"https://www.kickstarter.com/discover/places/nerja-nerja-es","location":"https://www.kickstarter.com/locations/nerja-nerja-es"},"api":{"nearby_projects":"https://api.kickstarter.com/v1/discover?signature=1576191145.87b54d70e4ee6e8256524d1403c85f7a5aefd29a&woe_id=768372"}}}'

- We've got information of the country, state and type of place in here

In [10]:
df['photo'][0]

'{"key":"assets/026/794/518/3c4061c549f67cbc846d085ccd24d906_original.jpg","full":"https://ksr-ugc.imgix.net/assets/026/794/518/3c4061c549f67cbc846d085ccd24d906_original.jpg?ixlib=rb-2.1.0&crop=faces&w=560&h=315&fit=crop&v=1570742981&auto=format&frame=1&q=92&s=c32fbce944b67b25c4c02b3483b9ce78","ed":"https://ksr-ugc.imgix.net/assets/026/794/518/3c4061c549f67cbc846d085ccd24d906_original.jpg?ixlib=rb-2.1.0&crop=faces&w=352&h=198&fit=crop&v=1570742981&auto=format&frame=1&q=92&s=53377f39b4fb3e57e024bd45563bb660","med":"https://ksr-ugc.imgix.net/assets/026/794/518/3c4061c549f67cbc846d085ccd24d906_original.jpg?ixlib=rb-2.1.0&crop=faces&w=272&h=153&fit=crop&v=1570742981&auto=format&frame=1&q=92&s=05cfdebf748a80173405bfc12721dd5b","little":"https://ksr-ugc.imgix.net/assets/026/794/518/3c4061c549f67cbc846d085ccd24d906_original.jpg?ixlib=rb-2.1.0&crop=faces&w=208&h=117&fit=crop&v=1570742981&auto=format&frame=1&q=92&s=1fc81fdb3fd4dd52b139267b861235f8","small":"https://ksr-ugc.imgix.net/assets/026/

- This columns does not have any useful info as image urls of many old projects do not work

In [11]:
df['profile'][0]

'{"id":3830311,"project_id":3830311,"state":"inactive","state_changed_at":1570742329,"name":null,"blurb":null,"background_color":null,"text_color":null,"link_background_color":null,"link_text_color":null,"link_text":null,"link_url":null,"show_feature_image":false,"background_image_opacity":0.8,"should_show_feature_image_section":true,"feature_image_attributes":{"image_urls":{"default":"https://ksr-ugc.imgix.net/assets/026/794/518/3c4061c549f67cbc846d085ccd24d906_original.jpg?ixlib=rb-2.1.0&crop=faces&w=1552&h=873&fit=crop&v=1570742981&auto=format&frame=1&q=92&s=79e2c06f403395d338c41f96cc8e91d4","baseball_card":"https://ksr-ugc.imgix.net/assets/026/794/518/3c4061c549f67cbc846d085ccd24d906_original.jpg?ixlib=rb-2.1.0&crop=faces&w=560&h=315&fit=crop&v=1570742981&auto=format&frame=1&q=92&s=c32fbce944b67b25c4c02b3483b9ce78"}}}'

- Here we have state_changed at which indicated when did the status of the projects change

In [12]:
df['urls'][0]

'{"web":{"project":"https://www.kickstarter.com/projects/organicfoods/organic-garden-boutique-hotel-in-sunny-spain?ref=discovery_category_newest","rewards":"https://www.kickstarter.com/projects/organicfoods/organic-garden-boutique-hotel-in-sunny-spain/rewards"}}'

- Project url leads us to the project home page
- rewards url leads us to the home page of the rewards section


**Let us extract category and sub category**

In [13]:
def get_cat(x):
    dic = eval(x)
    cat = dic['slug'].split("/")
    return cat[0]

def get_subcat(x):
    try:
        dic = eval(x)
        cat = dic['slug'].split("/")
        return cat[1]
    except:
        return "sub_"+str(cat[0])

In [14]:
df['cat'] = df['category'].apply(get_cat)
df['sub_cat'] = df['category'].apply(get_subcat)

- As we can see there are many stated project can exist, but given the context of our problem successful and failed are the only two relevant states so let us filter 

In [15]:
df['state'].value_counts()

successful    124272
failed         75960
canceled        8857
live            6063
suspended        648
Name: state, dtype: int64

In [16]:
def getValidState(x):
    try:
        if(x=='successful' or x=='failed'):
            return True
        else:
            return False
    except:
        return False

In [17]:
df = df[df['state'].apply(getValidState)]

In [18]:
df.isnull().mean()

Unnamed: 0                  0.000000
backers_count               0.000000
blurb                       0.000010
category                    0.000000
converted_pledged_amount    0.000000
country                     0.000000
country_displayable_name    0.000000
created_at                  0.000000
creator                     0.000000
currency                    0.000000
currency_symbol             0.000000
currency_trailing_code      0.000000
current_currency            0.000000
deadline                    0.000000
disable_communication       0.000000
friends                     0.999820
fx_rate                     0.000000
goal                        0.000000
id                          0.000000
is_backing                  0.999820
is_starrable                0.000000
is_starred                  0.999820
launched_at                 0.000000
location                    0.001034
name                        0.000000
permissions                 0.999820
photo                       0.000000
p

- We can confidently delete columns with 99% missing values
- at this point we can also remove columns out of which we have extracted information from

In [19]:
df.drop(['category', 'creator', 'profile', 'urls', 'photo', 'friends','is_backing', 'permissions', 'is_starred', 'location', 'source_url' ], axis=1 , inplace=True)

C:\Users\saima\Anaconda3\lib\site-packages\pandas\core\frame.py:4102: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


- We should remove duplicated to ensure the quality of the data

In [20]:
duplicate_bools = df['name'].duplicated()
df[duplicate_bools][['name' , 'blurb']]

name  \
604      WunderBar | The Farm-Fresh Vegan Clean Snack Bar   
2735              A minimalist verse novel, white on blue   
2921    Support Veterans with PTSD and promote a Trave...   
2966                                                Touch   
3495                   CUB Magazine - WWI Centenary Issue   
...                                                   ...   
215790  Dear Freshman Podcast - Advice to your freshma...   
215793  Styler Stories Presents #1: Tales of the Extra...   
215796  Photography Exhibition of Polaroid Manipulatio...   
215797                                 Reverie Cafe + Bar   
215798                              Introvert enamel pins   

                                                    blurb  
604     Delicious meets wholesome! WunderBar is full o...  
2735    John Johns: In perfect blue night a white owl ...  
2921    This project is being created to inspire and p...  
2966    A modern LGBT romance following several years ...  
3495    We are the award-winning Arts & Culture magazi...  
...                                                   ...  
215790  I will be creating a podcast interviewing acco...  
215793  Our premier comic book anthology, featuring 3 ...  
215796  An exhibition of my Polaroid Manipulations as ...  
215797  Plant Based, Big Taste: a scratch made kitchen...  
215798  A set of soft enamel pins for the introverts i...  

[25231 rows x 2 columns]

In [21]:
blurb_duplicate = set(df[df['blurb'].duplicated()].index.values)


name_duplicate = set(df[df['name'].duplicated()].index.values)




In [22]:
print(len(blurb_duplicate.symmetric_difference(name_duplicate)))

1666


- **Actually it took a lot of experimentation and coming back to this point of my project when i found out something did not add up(during model debugging) ,all that to find out how are duplicate values hidden the dataset**
- **at first glance it may seem easy just remove rows with all values as duplicate but that did not work out as some projects were logged multiple times in our data (due to some reason), which had different values some time dates and sometimes other feature(may be due to creator of that project updating the info**
- **Then i filtered duplicate values using just names and that also turned out to be incorrect in this case as multiple projects had same names, but it was too late as i already proceeded doing a lot of work based on this data(as you would see further down**
- **Finally blurb(Which is the description) is something i am considering to be unique and it is**

- **So looking at how many data points did we miss out assuming name to be unique is 1666 which is negligible compared to 2,00,00 rows we have**

In [23]:
df = df[~df['name'].duplicated()]


- Let us rename certain columns

In [24]:
df.rename(columns = {"country_displayable_name":"location_country", "state":"status", "cat":"category", "sub_cat":"sub_category"}, inplace=True)

In [25]:
df[:1].transpose()

1
Unnamed: 0                                                                1
backers_count                                                            59
blurb                     I need help! I'm recording my new record "Shad...
converted_pledged_amount                                               5818
country                                                                  US
location_country                                          the United States
created_at                                                       1355429246
currency                                                                USD
currency_symbol                                                           $
currency_trailing_code                                                 True
current_currency                                                        USD
deadline                                                         1363297878
disable_communication                                                 False
fx_rate                                                                   1
goal                                                                   5000
id                                                                455271642
is_starrable                                                          False
launched_at                                                      1360709478
name                                                          Shades Of Red
pledged                                                                5818
slug                                                          shades-of-red
spotlight                                                              True
staff_pick                                                            False
status                                                           successful
state_changed_at                                                 1363297878
static_usd_rate                                                           1
usd_pledged                                                            5818
usd_type                                                           domestic
category                                                              music
sub_category                                                            pop

- Country seems to simply be a duplicate for location_country(remove)
- currency_symbol isn't required because we already have currency(remove)
- Currency trailing code does not make sense (as in what it represents here)(remove)
- current_currency always seems to be the same as currency(remove)
- SLug seems to be the duplicate for name (remove)

In [26]:
drop_cols = ['country', 'currency_symbol', 'currency_trailing_code', 'current_currency', 'usd_type', 'slug', 'Unnamed: 0', 'id']

In [27]:
df.drop(drop_cols, axis=1, inplace=True)

In [28]:
df[:3].transpose()

1  \
backers_count                                                            59   
blurb                     I need help! I'm recording my new record "Shad...   
converted_pledged_amount                                               5818   
location_country                                          the United States   
created_at                                                       1355429246   
currency                                                                USD   
deadline                                                         1363297878   
disable_communication                                                 False   
fx_rate                                                                   1   
goal                                                                   5000   
is_starrable                                                          False   
launched_at                                                      1360709478   
name                                                          Shades Of Red   
pledged                                                                5818   
spotlight                                                              True   
staff_pick                                                            False   
status                                                           successful   
state_changed_at                                                 1363297878   
static_usd_rate                                                           1   
usd_pledged                                                            5818   
category                                                              music   
sub_category                                                            pop   

                                                                          2  \
backers_count                                                             0   
blurb                     The Perspective Project displays the art, the ...   
converted_pledged_amount                                                  0   
location_country                                         the United Kingdom   
created_at                                                       1518104045   
currency                                                                GBP   
deadline                                                         1523300074   
disable_communication                                                 False   
fx_rate                                                             1.32177   
goal                                                                   2000   
is_starrable                                                          False   
launched_at                                                      1518119674   
name                                                The Perspective Project   
pledged                                                                   0   
spotlight                                                             False   
staff_pick                                                            False   
status                                                               failed   
state_changed_at                                                 1523300074   
static_usd_rate                                                     1.39661   
usd_pledged                                                               0   
category                                                                art   
sub_category                                                    mixed media   

                                                                          3  
backers_count                                                           403  
blurb                           The Beloved "Learn To Draw" Series Returns!  
converted_pledged_amount                                              26733  
location_country                                          the United States  
created_at                                      

- Launched_at, deadline and state_changed seemed to be dates in the unix time let us convert them into date formats

In [29]:
from datetime import datetime
def get_date_format(x):
    ts = int(x)
    return (datetime.utcfromtimestamp(ts).strftime('%Y-%m-%d'))

In [30]:
df['launched_at'] = pd.to_datetime(df['launched_at'].apply(get_date_format))
df['deadline'] = pd.to_datetime(df['deadline'].apply(get_date_format))
df['state_changed_at'] = pd.to_datetime(df['state_changed_at'].apply(get_date_format))
df['created_at'] =  pd.to_datetime(df['created_at'].apply(get_date_format))

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 175001 entries, 1 to 215799
Data columns (total 22 columns):
backers_count               175001 non-null int64
blurb                       174999 non-null object
converted_pledged_amount    175001 non-null int64
location_country            175001 non-null object
created_at                  175001 non-null datetime64[ns]
currency                    175001 non-null object
deadline                    175001 non-null datetime64[ns]
disable_communication       175001 non-null bool
fx_rate                     175001 non-null float64
goal                        175001 non-null float64
is_starrable                175001 non-null bool
launched_at                 175001 non-null datetime64[ns]
name                        175001 non-null object
pledged                     175001 non-null float64
spotlight                   175001 non-null bool
staff_pick                  175001 non-null bool
status                      175001 non-null object
state

**We can save the dataset for further analysis**

In [30]:
df.to_csv("../data/notebooks/intial-prep.csv")